File: backends/dotnet/src/Zappy.Domain/Ordering/Totals.cs
namespace Zappy.Domain;

public sealed record Totals(Money Subtotal, Money Shipping, Money Discount, Money Total)
{
    public static readonly Money StandardShipping = Money.Euro(495);

    public static readonly Money FreeShippingFrom = Money.Euro(5000);

    public static Totals For(Money subtotal, PromotionRule? promotionRule)
    {
        var nothing = Money.ZeroIn(subtotal.Currency);
        var discount = promotionRule is null ? nothing : promotionRule.DiscountFor(subtotal).CappedAt(subtotal);
        var charged = subtotal.IsZero || subtotal.IsAtLeast(FreeShippingFrom) ? nothing : StandardShipping;
        var shipping = promotionRule is null ? charged : promotionRule.ShippingFor(charged);
        return new Totals(subtotal, shipping, discount, subtotal.Plus(shipping).Minus(discount));
    }
}

File: backends/node/subgraphs/promotions/src/domain/totals.ts
import type { Money } from "@zappy/shared";
import { addMoney, money, subtractMoney, zeroMoney } from "@zappy/shared";

export const shippingCharge = money(495);

export const freeShippingThreshold = money(5000);

export function shippingFor(subtotal: Money, freeShippingApplies: boolean): Money {
  if (subtotal.amount === 0) {
    return zeroMoney;
  }
  if (freeShippingApplies) {
    return zeroMoney;
  }
  return subtotal.amount >= freeShippingThreshold.amount ? zeroMoney : shippingCharge;
}

export function totalFor(subtotal: Money, shipping: Money, discount: Money): Money {
  return subtractMoney(addMoney(subtotal, shipping), discount);
}
